### PR TITLE
Add hardware loop code generation

### PR DESCRIPTION
**Pull request by serkm**
We use the `HardwareLoops` pass before instruction selection to insert [hardware loop intrinsics](https://llvm.org/docs/LangRef.html#hardware-loop-intrinsics) into LLVM-IR. This pass calls `isHardwareLoopProfitable` which checks [loops](https://llvm.org/docs/LoopTerminology.html) for following constraints.
- The `xcorevhwlp` extension is enabled
- The loop needs exactly one basic block as latch and exiting block
- The trip count needs to be loop invariant and representable in a register
- The loop contains no instruction that might get lowered to a branch instruction
- The control flow graph of the loop body needs to be linear (except for nested hardware loops)

The intrinsics `@llvm.set.loop.iterations.i32` and `@llvm.loop.decrement.i32` with a conditional branch are lowered to the pseudo instructions `HwlpSetup`/`HwlpSetupImm` and `HwlpBranch`, respectively.

The `RISCVCoreVHwlpBlocks` pass is inserted before register allocation. This pass reorders the blocks of hardware loops to eliminate branch instructions and adds the blocks to `RISCVMachineFunctionInfo` to prevent analysing their terminators and to disable instruction compression. (reusing [this commit](https://github.com/flip1995/corev-llvm-project/commit/90fd0a64dab0bccc17d23ea051d186434ca4bbb2)).

Before emitting the machine code, the `RISCVExpandCoreVHwlpPseudo` pass replaces the pseudo instructions with the proper instructions from `xcorevhwlp`. We have to do this as late as possible to make sure that the last instruction in the loop won't be moved.

### CODE REVIEW

**Comment by flip1995** [comment]
Thanks for the PR! I started reviewing this but will have to spend a little more time for a thorough review. After skimming the code, the approaches taken here seem to be way better and more stable than the hacks I attempted in #3.

File: llvm/lib/Target/RISCV/RISCVExpandCoreVHwlpPseudoInsts.cpp [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        break;
+      default:
+        break;
+    }
+    if (InnerEndSymbol && InnerEndSymbol == MI->getPreInstrSymbol()) {
+      InnerHwlpEndOffset = Offset;
+    }
+    unsigned Size = TII->getInstSizeInBytes(*MI);
+    // Instruction compression is disabled here
+    if (Size == 2){
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVExpandCoreVHwlpPseudoInsts.cpp:144]: Isn't this just an error? How can an instruction of size 2 exist in a HWLP?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 2]
```diff
@@ -98,3 +98,119 @@ int RISCVTTIImpl::getIntImmCostIntrin(Intrinsic::ID IID, unsigned Idx,
...
   return ST->hasExtXCoreVMem();
 }
+
+bool RISCVTTIImpl::isHardwareLoopProfitable(Loop *L, ScalarEvolution &SE,
+                                            AssumptionCache &AC,
+                                            TargetLibraryInfo *LibInfo,
+                                            HardwareLoopInfo &HWLoopInfo) {
+  if (!ST->hasExtXCoreVHwlp()) {
+    return false;
+  }
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: NIT usually those are written without braces:

```suggestion
  if (!ST->hasExtXCoreVHwlp())
    return false;
```

Same with the once below.

**Review by flip1995 — COMMENTED** [review]
Sorry for the long time this took me to review. The last month was really busy for me. 

You marked this PR as "WIP". Do you have any specific plans on things you want to add to this PR before it is ready?

File: llvm/lib/Target/RISCV/RISCVExpandCoreVHwlpPseudoInsts.cpp (hunk 1 above)
Comment (serkm) [reply llvm/lib/Target/RISCV/RISCVExpandCoreVHwlpPseudoInsts.cpp:144]: `getInstSizeInBytes` doesn't know that instruction compression is not allowed here and returns 2 on compressible instructions.

**Comment by serkm** [comment]
I must have forgotten to remove the WIP tag. I don't have anything to add to this PR.

**Comment by jeremybennett** [comment]
@flip1995 Is this good to merge now?

**Comment by flip1995** [comment]
I think so. I'll take a final look tomorrow and will ping you if it can get merged.

**Comment by flip1995** [comment]
This LGTM now. @jeremybennett ready to merge.

**Comment by jeremybennett** [comment]
All merged - thanks @SerkanMuhcu and @flip1995

**Pull request merged**